File: pset6/credit.py
import cs50

def main():
    #user input
    print("Number: ", end="")
    card_no = cs50.get_int()
    counter = 0
    checksum = 0
    prefix = 0
    prefix_switch = False
    card_supplier = "INVALID"
    
    #iterate through digits by modulo
    while (card_no > 0):
        
        #strip last digit
        counter += 1
        last_digit = card_no % 10
        
        #checksum
        if counter % 2 == 0:
            last_digit *= 2
            if last_digit >= 10:
                checksum += 1 + (last_digit - 10)
            else:
                checksum += last_digit
                
        else:
            checksum += last_digit
            
        #return card number, minus last digit
        card_no = card_no // 10
        
        #supplier check
        if card_no < 99 and prefix_switch == False:
            prefix_switch = True
            
            if (card_no // 10 == 4) and (counter == 14 or 11):
                card_supplier = "VISA"
                
            elif (card_no == 51 or 52 or 53 or 54 or 55) and counter == 14:
                card_supplier = "MASTERCARD"
    
            elif (card_no == 34 or 37) and counter == 13:
                card_supplier = "AMEX"
                
    #reject card_no with wrong checksum
    if checksum % 10 != 0:
        card_supplier = "INVALID"
        
    #output
    print(card_supplier)

if __name__ == "__main__":
    main()

File: pset1/credit.c
#include <stdio.h>
#include <cs50.h>

int main(void)
{
    printf("Number: ");
    long long cardNumber = get_long_long();
    long long cardNumberCopy = cardNumber;
    int digit = 0;
    int checksum = 0;
    int firstTwoDigits;
    
    while (cardNumberCopy > 0)
    {
        if (cardNumberCopy < 100 && cardNumberCopy > 9)
        {
            firstTwoDigits = cardNumberCopy;
        }
        int lastdigit = cardNumberCopy % 10;
        digit++;
        
        if (digit % 2 == 0)
        {
            if (lastdigit >= 5)
            {
                checksum += 1 + (lastdigit * 2 - 10);
            }
            else
            {
                checksum += lastdigit * 2;
                }
        }
        else
        {
            checksum += lastdigit;
        }
        cardNumberCopy /= 10;
    }
    
    if (checksum % 10 != 0)
    {
        printf("INVALID\n");
    }
    else if (digit == 13)
    {
        switch (firstTwoDigits/10)
        {
            case 4:
                printf("VISA\n");
                break;
            default:
                printf("INVALID\n");
                break;
        }
    }
    else if (digit == 15)
    {
        switch (firstTwoDigits)
        {
            case 34:
            case 37:
                printf("AMEX\n");
                break;
            default:
                printf("INVALID\n");
                break;
        }
    }
    else if (digit == 16)
    {
        switch (firstTwoDigits)
        {
            case 40:
            case 41:
            case 42:
            case 43:
            case 44:
            case 45:
            case 46:
            case 47:
            case 48:
            case 49:
                printf("VISA\n");
                break;
            case 51:
            case 52:
            case 53:
            case 54:
            case 55:
                printf("MASTERCARD\n");
                break;
            default:
                printf("INVALID\n");
                break;
        }
    }
    return 0;
}